### ISSUE TITLE
Examplar task results reproduction

### ISSUE THREAD

**Issue by yohann84L** [nupurkmr9/S2M2_fewshot#7]
I've been trough your paper (which is excellent by the way!) and you have results with 2 differents pretext tasks (rotation & examplar). 

I did not found in your code the part for examplar task. Could you explain a bit more how did you use it? Code example?Which augmentation did you use?

Thanks a lot!

**Comment by nupurkmr9** [comment]
Hi, 

Sorry for the late response. 
The repo doesn't include code for the exemplar pretext task since we got best results with manifold mixup over rotation pretext task. 
In case of exemplar training we use the same transformations as defined in datamgr.py : _['RandomSizedCrop', 'ImageJitter', 'RandomHorizontalFlip'_] and create 4 exemplars for each image. 

During exemplar training, we perform hard negative mining and use _nn.SoftMarginLoss()_  along with _nn.CrossEntropyLoss()_

Code for exemplar loss: 

`

    def euclidean_dist(x, y):
        m, n = x.size(0), y.size(0)
        xx = torch.pow(x, 2).sum(1, keepdim=True).expand(m, n)
        yy = torch.pow(y, 2).sum(1, keepdim=True).expand(n, m).t()
        dist = xx + yy
        dist.addmm_(1, -2, x, y.t())
        dist = dist.clamp(min=1e-12).sqrt()  # for numerical stability
        return dist

    def hard_example_mining(dist_mat, labels, return_inds=False):
        N = dist_mat.size(0)  # shape [N, N]
        is_pos = labels.expand(N, N).eq(labels.expand(N, N).t())
        is_neg = labels.expand(N, N).ne(labels.expand(N, N).t())
        dist_ap, relative_p_inds = torch.max(
            dist_mat[is_pos].contiguous().view(N, -1), 1, keepdim=True) #distance(anchor, positive)
        dist_an, relative_n_inds = torch.min(
            dist_mat[is_neg].contiguous().view(N, -1), 1, keepdim=True) # distance(anchor, negative)
        dist_ap = dist_ap.squeeze(1)
        dist_an = dist_an.squeeze(1)

        if return_inds:
            ind = (labels.new().resize_as_(labels)
                   .copy_(torch.arange(0, N).long())
                   .unsqueeze(0).expand(N, N))
            p_inds = torch.gather(
                ind[is_pos].contiguous().view(N, -1), 1, relative_p_inds.data)
            n_inds = torch.gather(
                ind[is_neg].contiguous().view(N, -1), 1, relative_n_inds.data)
            p_inds = p_inds.squeeze(1)
            n_inds = n_inds.squeeze(1)
            return dist_ap, dist_an, p_inds, n_inds

        return dist_ap, dist_an

    ranking_loss = nn.SoftMarginLoss()
    def exemplar_loss_fn(f, a) :

        dist_mat = euclidean_dist(f, f)
        dist_ap, dist_an = hard_example_mining(dist_mat, a, return_inds=False) # a is exemplar labels 
        y = dist_an.new().resize_as_(dist_an).fill_(1)
        loss = ranking_loss(dist_an - dist_ap, y)
        return loss

`

**Comment by yohann84L** [comment]
Great ! Thanks a lot for your answer I'll look into it :)
